lambdas/harvesters/ror/downloader.rb

```diff
@@ -0,0 +1,355 @@
+        resp = client.get_item({ table_name: table, key:, consistent_read: false,
+                                 return_consumed_capacity: logger&.level == 'debug' ? 'TOTAL' : 'NONE' })
+
+        logger.debug(message: "#{SOURCE} fetched DMP ID: #{key}") if logger.respond_to?(:debug)
+        resp[:item].is_a?(Array) ? resp[:item].first : resp[:item]
+      rescue Aws::Errors::ServiceError => e
+        logger&.error(message: format(MSG_DYNAMO_ERROR, msg: e.message, trace: e.backtrace))
+        nil
+      end
+
+      # Update the harvester's record in the RDS database
+      def update_harvester_record(client:, table:, source:, metadata: nil, tstamp: nil, logger: nil)
+        source['last_metadata'] = metadata
+        source['last_synced_at'] = tstamp
+        resp = client.put_item({ table_name: table, item: source,
+                                 return_consumed_capacity: logger&.level == 'debug' ? 'TOTAL' : 'NONE' })
+
+        logger.debug(message: "#{SOURCE} put_item HARVESTER: #{source['ID']}", details: source) if logger.respond_to?(:debug)
+        resp
+      rescue Aws::Errors::ServiceError => e
+        logger&.error(message: format(MSG_DYNAMO_ERROR, msg: e.message, trace: e.backtrace))
+        nil
+      end
+
+      # Fetch the ROR Harvester record from the RDS instance or create it
+      def find_or_create_source(client:, table:, logger: nil)
+        record = fetch_harvester_record(client:, table:, logger:)
+        return record unless record.nil?
+
+        # If it was not found, create the record and then return the new record
+        source = {
+          RESOURCE_TYPE: HARVESTER_PK,
+          ID: HARVESTER_SK,
+          name: 'ROR Harvester',
+          _TYPE: 'download'
+        }
+        update_harvester_record(client:, table:, source:)
+        fetch_harvester_record(client:, table:, logger:)
+      rescue StandardError => e
+        logger&.error(message: "ERROR when trying to create a ROR harvester record in RDS: #{e.message}", details: e.backtrace)
+        nil
+      end
+
+      # Fetch the latest ROR archive file metadata from Zenodo
+      def fetch_zenodo_metadata(logger: nil)
+        # Fetch the latest ROR metadata from Zenodo (the query will place the most recent version 1st)
+        resp = Uc3DmpExternalApi::Client.call(url: ZENODO_METADATA_TARGET, method: :get,
+                                              additional_headers: { host: HEADERS_HOST })
+
+        # Extract the most recent file's metadata
+        metadata = resp.fetch('hits', {}).fetch('hits', []).first&.fetch('files', [])&.last
+        logger&.error(message: 'Unable to download ROR file metadata from Zenodo') if metadata.nil?
+        url = metadata.fetch('links', {}).fetch('download', metadata.fetch('links', {})['self'])
+        logger&.error(message: 'Zenodo file metadata is missing file download URL.') if url.nil?
+        metadata[:download_url] = url
+        JSON.parse(metadata.to_json)
```

Comment:
L162: Does "JSON.parse()" and ".to_json" negate one another? It seems like you are converting the metadata to JSON, but then parsing it back into a hash.

Reply:
It would seem that way, but ruby has a data type called Symbol which creates issues. You reference them as `hash[:key]` instead of `hash['key']`. Ruby uses Symbol by default even if you do `hash['key'] = 'value'`. Its super annoying.

That means you can end up with things like: `hash = { a: 'my key is a symbol', 'b': 'my key is a string'}` and then when you try to access them with `hash['a']` it won't find it 🙃 

Calling `.to_json` converts the Hash/Array into a string and then `JSON.parse` converts it back into a Hash/Array but with all of the keys as Strings.

Reply:
Oh that's good to know. Thanks for the info Brian.